tpm2/keydata_v3.go

```diff
@@ -0,0 +1,136 @@
+	d.KeyImportSymSeed = nil
+}
+
+func (d *keyData_v3) ValidateData(tpm *tpm2.TPMContext, session tpm2.SessionContext) (tpm2.ResourceContext, error) {
+	if d.KeyImportSymSeed != nil {
+		return nil, errors.New("cannot validate importable key data")
+	}
+
+	// Validate the type and scheme of the dynamic authorization policy signing key.
+	authPublicKey := d.PolicyData.StaticData.AuthPublicKey
+	authKeyName, err := authPublicKey.Name()
+	if err != nil {
+		return nil, keyDataError{xerrors.Errorf("cannot compute name of dynamic authorization policy key: %w", err)}
+	}
+	if authPublicKey.Type != tpm2.ObjectTypeECC {
+		return nil, keyDataError{errors.New("public area of dynamic authorization policy signing key has the wrong type")}
+	}
+	authKeyScheme := authPublicKey.Params.AsymDetail(authPublicKey.Type).Scheme
+	if authKeyScheme.Scheme != tpm2.AsymSchemeNull {
+		if authKeyScheme.Scheme != tpm2.AsymSchemeECDSA {
+			return nil, keyDataError{errors.New("dynamic authorization policy signing key has unexpected scheme")}
+		}
+		if authKeyScheme.Details.Any(authKeyScheme.Scheme).HashAlg != authPublicKey.NameAlg {
+			return nil, keyDataError{errors.New("dynamic authorization policy signing key algorithm must match name algorithm")}
+		}
+	}
+
+	// Create a context for the PCR policy counter.
+	pcrPolicyCounterHandle := d.PolicyData.StaticData.PCRPolicyCounterHandle
+	var pcrPolicyCounter tpm2.ResourceContext
+	switch {
+	case pcrPolicyCounterHandle != tpm2.HandleNull && pcrPolicyCounterHandle.Type() != tpm2.HandleTypeNVIndex:
+		return nil, keyDataError{errors.New("PCR policy counter handle is invalid")}
+	case pcrPolicyCounterHandle != tpm2.HandleNull:
+		pcrPolicyCounter, err = tpm.CreateResourceContextFromTPM(pcrPolicyCounterHandle, session.IncludeAttrs(tpm2.AttrAudit))
+		if err != nil {
+			if tpm2.IsResourceUnavailableError(err, pcrPolicyCounterHandle) {
+				return nil, keyDataError{errors.New("PCR policy counter is unavailable")}
+			}
+			return nil, xerrors.Errorf("cannot create context for PCR policy counter: %w", err)
+		}
+	}
+
+	// Make sure that the static authorization policy data is consistent with the sealed key object's policy.
+	if !d.KeyPublic.NameAlg.Available() {
+		return nil, keyDataError{errors.New("cannot determine if static authorization policy matches sealed key object: algorithm unavailable")}
+	}
+	trial := util.ComputeAuthPolicy(d.KeyPublic.NameAlg)
+	trial.PolicyAuthorize(computeV3PcrPolicyRefFromCounterContext(pcrPolicyCounter), authKeyName)
+	trial.PolicyAuthValue()
+
+	if !bytes.Equal(trial.GetDigest(), d.KeyPublic.AuthPolicy) {
+		return nil, keyDataError{errors.New("the sealed key object's authorization policy is inconsistent with the associated metadata or persistent TPM resources")}
+	}
+
+	return pcrPolicyCounter, nil
+}
+
+func (d *keyData_v3) Write(w io.Writer) error {
+	_, err := mu.MarshalToWriter(w, d)
+	return err
+}
+
+func (d *keyData_v3) Policy() keyDataPolicy {
+	return d.PolicyData
```

Comment:
this is not used/covered by tests

Reply:
It looks like there's no explicit coverage for this in previous versions either - it gets used by other code. That said, it looks like the extra tests I've added for ValidateData use this now.